dbms/src/Common/Spiller.cpp

```diff
@@ -0,0 +1,148 @@
+    {
+        spill_dir += Poco::Path::separator();
+    }
+    for (size_t i = 0; i < partition_num; ++i)
+        spilled_files.push_back(std::make_unique<SpilledFiles>());
+}
+
+bool Spiller::spillBlocks(const Blocks & blocks, size_t partition_id)
+{
+    RUNTIME_CHECK_MSG(partition_id < partition_num, "{}: partition id {} exceeds partition num {}.", id, partition_id, partition_num);
+    RUNTIME_CHECK_MSG(spill_finished == false, "{}: spill after the spiller is finished.", id);
+    /// todo append to existing file
+    if (unlikely(blocks.empty()))
+        return true;
+    auto spilled_file_name = nextSpillFileName(partition_id);
+    try
+    {
+        auto spilled_file = std::make_unique<SpilledFile>(spilled_file_name);
+        RUNTIME_CHECK_MSG(!spilled_file->exists(), "Duplicated spilled file: {}, should not happens", spilled_file_name);
+        WriteBufferFromFile file_buf(spilled_file_name);
+        CompressedWriteBuffer compressed_buf(file_buf);
+        NativeBlockOutputStream block_out(compressed_buf, 0, blocks[0].cloneEmpty());
+        block_out.writePrefix();
+        for (const auto & block : blocks)
+        {
+            auto block_bytes_size = block.bytes();
+            block_out.write(block);
+            spilled_file->addSpilledDataSize(block_bytes_size);
+        }
+        {
+            std::lock_guard lock(spilled_files[partition_id]->spilled_files_mutex);
+            spilled_files[partition_id]->spilled_files.emplace_back(std::move(spilled_file));
+        }
+        return true;
+    }
+    catch (...)
+    {
+        LOG_ERROR(logger, "Failed to spill block to disk for file {}, error message: {}", spilled_file_name, getCurrentExceptionMessage(false, false));
+        return false;
+    }
+}
+
+BlockInputStreams Spiller::restoreBlocks(size_t partition_id, size_t max_stream_size)
+{
+    RUNTIME_CHECK_MSG(partition_id < partition_num, "{}: partition id {} exceeds partition num {}.", id, partition_id, partition_num);
+    RUNTIME_CHECK_MSG(spill_finished, "{}: restore before the spiller is finished.", id);
+    if (is_input_sorted && spilled_files[partition_id]->spilled_files.size() > max_stream_size)
+        LOG_WARNING(logger, "sorted spilled data restore does not take max_stream_size into account");
+    BlockInputStreams ret;
+    if (is_input_sorted)
+    {
+        for (const auto & file : spilled_files[partition_id]->spilled_files)
+        {
+            RUNTIME_CHECK_MSG(file->exists(), "Spill file {} does not exists", file->path());
+            std::vector<String> files{file->path()};
+            ret.push_back(std::make_shared<SpilledFilesInputStream>(files, input_schema));
+        }
+    }
+    else
+    {
+        size_t return_stream_num = std::min(max_stream_size, spilled_files[partition_id]->spilled_files.size());
+        std::vector<std::vector<String>> files(return_stream_num);
+        for (size_t i = 0; i < spilled_files[partition_id]->spilled_files.size(); ++i)
```

Comment:
Maybe we can add todo here.
`TODO balance based on SpilledDataSize`

Reply:
Done